### PR TITLE
resolves #6, Added `KmsAsymmetricSigner` and `KmsAsymmetricVerifier` which supports both RSA and ECDSA algorithms.

### LINKED ISSUE #6: Add support for ECDSA algorithms for asymmetric singer and verifiers

### PR DESCRIPTION
**Pull request by debanshuk**
Added `KmsAsymmetricSigner` and `KmsAsymmetricVerifier` which supports both RSA and ECDSA algorithms. Also deprecated `KmsAsymmetricRSASSASigner` and `KmsAsymmetricRSASSAVerifier`, while retaining them for backwards compatibility.

Few Notes: 
1. The newly added `KmsAsymmetricSigner` and `KmsAsymmetricVerifier` classes have the exact same code as `KmsAsymmetricRSASSASigner` and `KmsAsymmetricRSASSAVerifier` respectively. This is because the code for RSA and ECDSA is exactly the same. [Test cases are also the same.]
1. Similar to point #1, the `KmsAsymmetricSigningCryptoProvider` class the exact same code as `KmsAsymmetricRSASSAProvider` class, with addition of ECDSA algorithms in the support algorithms. [Test cases are also the same.]
1. Deprecated the older `KmsAsymmetricRSASSASigner`, `KmsAsymmetricRSASSAVerifier` and `KmsAsymmetricRSASSAProvider` classes, while keeping them for backwards compatibility.
1. I have added in the 'SM2DSA' signing algorithm, since it uses 'SM3' hashing algorithm, which is not supported by hashing lib we are using. Addition of this algorithm should be taken in a separate PR.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: nimbus-jose-jwt_aws-kms-extension/src/main/java/com/nimbusds/jose/aws/kms/crypto/impl/KmsAsymmetricSigningCryptoProvider.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        this.messageType = messageType;
+    }
+
+    protected ByteBuffer getMessage(final JWSHeader header, final byte[] signingInput) throws JOSEException {
+        final var alg = header.getAlgorithm();
+        var message = signingInput;
+
+        String messageDigestAlgorithm = Optional.ofNullable(JWS_ALGORITHM_TO_MESSAGE_DIGEST_ALGORITHM.get(alg))
+                .orElseThrow(() -> new JOSEException(
+                        String.format("No digest algorithm exist for JWE algorithm %s in map: %s",
```
Comment (mjsheikh) [inline nimbus-jose-jwt_aws-kms-extension/src/main/java/com/nimbusds/jose/aws/kms/crypto/impl/KmsAsymmetricSigningCryptoProvider.java:111]: This should rather be JWS algorithm. (JWE mentioned)

File: nimbus-jose-jwt_aws-kms-extension/src/main/java/com/nimbusds/jose/aws/kms/crypto/KmsAsymmetricVerifier.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import com.nimbusds.jose.crypto.impl.CriticalHeaderParamsDeferral;
+import com.nimbusds.jose.util.Base64URL;
+import java.nio.ByteBuffer;
+import java.util.Set;
+import javax.annotation.concurrent.ThreadSafe;
+import lombok.NonNull;
+import lombok.var;
+
+/**
+ * Sign verifier implementation for RSA-SSA signing with public/private key stored in AWS KMS.
```
Comment (mjsheikh) [inline nimbus-jose-jwt_aws-kms-extension/src/main/java/com/nimbusds/jose/aws/kms/crypto/KmsAsymmetricVerifier.java:49]: [Nit] Same as above. Kindly add supported algorithms in comment.

**Review by AnupamAS0x1 — APPROVED** [review]
looks good

File: nimbus-jose-jwt_aws-kms-extension/src/main/java/com/nimbusds/jose/aws/kms/crypto/impl/KmsAsymmetricSigningCryptoProvider.java (hunk 1 above)
Comment (debanshuk) [reply nimbus-jose-jwt_aws-kms-extension/src/main/java/com/nimbusds/jose/aws/kms/crypto/impl/KmsAsymmetricSigningCryptoProvider.java:111]: Fixed.

File: nimbus-jose-jwt_aws-kms-extension/src/main/java/com/nimbusds/jose/aws/kms/crypto/KmsAsymmetricVerifier.java (hunk 2 above)
Comment (debanshuk) [reply nimbus-jose-jwt_aws-kms-extension/src/main/java/com/nimbusds/jose/aws/kms/crypto/KmsAsymmetricVerifier.java:49]: Fixed.

File: nimbus-jose-jwt_aws-kms-extension/src/main/java/com/nimbusds/jose/aws/kms/crypto/KmsAsymmetricSigner.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import com.nimbusds.jose.aws.kms.crypto.impl.KmsAsymmetricSigningCryptoProvider;
+import com.nimbusds.jose.aws.kms.exceptions.TemporaryJOSEException;
+import com.nimbusds.jose.util.Base64URL;
+import javax.annotation.concurrent.ThreadSafe;
+import lombok.NonNull;
+import lombok.var;
+
+
+/**
+ * Signer implementation for RSA-SSA signing with public/private key stored in AWS KMS.
```
[the remark this answers is not in the document]
Comment (debanshuk) [reply nimbus-jose-jwt_aws-kms-extension/src/main/java/com/nimbusds/jose/aws/kms/crypto/KmsAsymmetricSigner.java:45]: Fixed.

File: nimbus-jose-jwt_aws-kms-extension/src/main/java/com/nimbusds/jose/aws/kms/crypto/impl/KmsAsymmetricSigningCryptoProvider.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+        if (messageType == MessageType.DIGEST) {
+            MessageDigest messageDigestProvider;
+            try {
+                messageDigestProvider = MessageDigest.getInstance(messageDigestAlgorithm);
+            } catch (NoSuchAlgorithmException e) {
+                throw new JOSEException("Invalid message digest algorithm.", e);
+            }
+            message = messageDigestProvider.digest(message);
+        }
```
Comment (mjsheikh) [inline nimbus-jose-jwt_aws-kms-extension/src/main/java/com/nimbusds/jose/aws/kms/crypto/impl/KmsAsymmetricSigningCryptoProvider.java:122]: [Might be taken as follow up re-work]
This behaviour doesn't seem ideal to me from a user's perspective. 
Internally I don't believe there is any difference in the way hashing is done, either by `messageDigestProvider` or hashing done by KMS. Why would clients care at which layer the hashing is being done. They can only pass RAW unhashed inputs here. 

Wouldn't it make more sense to rather allow them to pass both kinds of messages (`RAW`/`DIGEST`) and then just pass that to KMS?

Comment (mjsheikh) [reply nimbus-jose-jwt_aws-kms-extension/src/main/java/com/nimbusds/jose/aws/kms/crypto/impl/KmsAsymmetricSigningCryptoProvider.java:122]: One counter argument can be that we'll rely on clients to insure that the passed messageType is correct. Can't we have validations for such a case?

**Review by mjsheikh — APPROVED** [review]

**Pull request merged**